app/routers/menstrual_predictor.py

```diff
@@ -0,0 +1,179 @@
+
+
+router = APIRouter(
+    prefix="/menstrual_predictor",
+    tags=["menstrual_predictor"],
+    dependencies=[Depends(get_db)]
+)
+
+MENSTRUAL_PERIOD_CATEGORY_ID = 111
+
+
+@router.get("/")
+def join(request: Request, db: Session = Depends(get_db)):
+    current_user_id = get_current_user(db).id
+
+    if not is_user_signed_up_to_menstrual_predictor(db, current_user_id):
+        logger.info('getting menstrual predictor')
+        return templates.TemplateResponse("join_menstrual_predictor.html", {
+            "request": request,
+        })
+    else:
+        return RedirectResponse(url='/', status_code=HTTP_302_FOUND)
+
+
+@router.get("/add-period-start/{start_date}")
+def add_period_start(request: Request, start_date: str,
+                     db: Session = Depends(get_db)):
+    try:
+        period_start_date = datetime.datetime.strptime(start_date, '%Y-%m-%d')
+    except ValueError as err:
+        logger.exception(err)
+        raise HTTPException(
+            status_code=400,
+            detail="The given date doesn't match a date format YYYY-MM-DD")
+    else:
+        current_user_id = get_current_user(db).id
+        user_period_length = is_user_signed_up_to_menstrual_predictor(
+            db, current_user_id)
+
+        remove_existing_period_dates(db, current_user_id)
+        if user_period_length:
+            add_3_month_predictions(
+                db, user_period_length, period_start_date, current_user_id)
+    logger.info('adding menstrual start date')
+    return RedirectResponse('/', status_code=303)
+
+
+@router.post("/")
+async def submit_join_form(
+    request: Request,
+    db: Session = Depends(get_db)
+):
+
+    data = await request.form()
+    current_user = get_current_user(session=db)
+
+    user_menstrual_period_length = {
+        "user_id": current_user.id,
+        "period_length": data['avg_period_length'],
+    }
+    last_period_date = datetime.datetime.strptime(
+        data['last_period_date'], '%Y-%m-%d')
+    try:
+        new_signed_up_to_period = create_model(
+            session=db,
+            model_class=UserMenstrualPeriodLength,
+            **user_menstrual_period_length)
+    except SQLAlchemyError:
+        logger.info(
+            f'User {new_signed_up_to_period.user_id}'
+            'already signed up to the service, hurray')
+        db.rollback()
+    url = '/'
+    generate_predicted_period_dates(
+        db, data['avg_period_length'], last_period_date, current_user.id)
+
+    return RedirectResponse(url=url, status_code=HTTP_302_FOUND)
+
+
+@router.get("/get-period-dates/")
+def get_period_days(request: Request, db: Session = Depends(get_db)):
+    period_days = get_all_period_days(db, 1)
+    gaps_list = []
+    for i in range(len(period_days) - 1):
+        gap = get_date_diff(period_days[i].start, period_days[i + 1].start)
+        gaps_list.append(gap.days)
+    logger.critical(get_list_avg(gaps_list))
+
+
+def get_avg_period_gap(db: Session, user_id):
```

Comment:
From here to the bottom, I think it's better to have the functions inside `/internal`

Reply:
done